libs/math/tests/math/all_tests/clustering/knn_test.cpp

```diff
@@ -84,41 +57,13 @@ TYPED_TEST(ClusteringTest, knn_euclidean_test)
 
 TYPED_TEST(ClusteringTest, knn_cosine_test)
 {
-  using DataType  = typename TypeParam::Type;
   using ArrayType = TypeParam;
   using SizeType  = typename TypeParam::SizeType;
 
-  ArrayType A = ArrayType({4, 4});
-
-  A.Set(SizeType{0}, SizeType{0}, DataType(1));
-  A.Set(SizeType{0}, SizeType{1}, DataType(2));
-  A.Set(SizeType{0}, SizeType{2}, DataType(3));
-  A.Set(SizeType{0}, SizeType{3}, DataType(4));
-
-  A.Set(SizeType{1}, SizeType{0}, DataType(2));
-  A.Set(SizeType{1}, SizeType{1}, DataType(3));
-  A.Set(SizeType{1}, SizeType{2}, DataType(4));
-  A.Set(SizeType{1}, SizeType{3}, DataType(5));
-
-  A.Set(SizeType{2}, SizeType{0}, DataType(-1));
-  A.Set(SizeType{2}, SizeType{1}, DataType(-2));
-  A.Set(SizeType{2}, SizeType{2}, DataType(-3));
-  A.Set(SizeType{2}, SizeType{3}, DataType(-4));
-
-  A.Set(SizeType{3}, SizeType{0}, DataType(-2));
-  A.Set(SizeType{3}, SizeType{1}, DataType(-3));
-  A.Set(SizeType{3}, SizeType{2}, DataType(-4));
-  A.Set(SizeType{3}, SizeType{3}, DataType(-5));
-
-  ArrayType v = ArrayType({1, 4});
-
-  v.Set(SizeType{0}, SizeType{0}, DataType(3));
-  v.Set(SizeType{0}, SizeType{1}, DataType(4));
-  v.Set(SizeType{0}, SizeType{2}, DataType(5));
-  v.Set(SizeType{0}, SizeType{3}, DataType(6));
+  ArrayType A = ArrayType::FromString("1, 2, 3, 4; 2, 3, 4, 5; -1, -2, -3, -4; -2, -3, -4, -5");
+  ArrayType v = ArrayType::FromString("3, 4, 5, 6");
 
-  std::vector<std::pair<typename ArrayType::SizeType, typename ArrayType::Type>> output =
-      fetch::math::clustering::KNNCosine(A, v, 4);
+  auto output = fetch::math::clustering::KNNCosine(A, v, 4);
```

Comment:
I don't want to be too pedantic, but you are using namespace fetch::math::clustering;